src/dns.rs

```diff
@@ -0,0 +1,305 @@
+use core::mem;
+
+use anyhow::*;
+use log::*;
+
+use esp_idf_sys::c_types::*;
+use esp_idf_sys::*;
+
+use crate::private::cstr::{CStr, CString};
+use crate::task::{TaskConfig, TaskHandle};
+use crate::lwip;
+
+/// 0.0.0.0
+const IPADDR_ANY: u32 = 0x00000000;
+
+const DNS_PORT: c_ushort = 53;
+const DNS_MAX_LEN: usize = 256;
+
+const OPCODE_MASK: u16 = 0x7800;
+const QR_FLAG: u16 = 1 << 7;
+const QD_TYPE_A: u16 = 0x0001;
+const ANS_TTL_SEC: u32 = 300;
+
+pub struct CaptivePortalDns {
+    task_handle: Option<TaskHandle>,
+}
+
+impl CaptivePortalDns {
+    pub fn new() -> Self {
+        CaptivePortalDns { task_handle: None }
```

Comment:
I guess the desire to stay `no_std` compatible is what  brought you into exposing bits and pieces for FreeRTOS? How about we take a simple trait instead, that would allow us to spawn a thread? We can even provide an implementation of the trait when Rust STD is enabled, and provide additional constructor which does not take the trait, for STD mode.

Reply:
> I guess the desire to stay `no_std` compatible is what brought you into exposing bits and pieces for FreeRTOS?
Yes

> How about we take a simple trait instead, that would allow us to spawn a thread? We can even provide an implementation of the trait when Rust STD is enabled, and provide additional constructor which does not take the trait, for STD mode.

As in, implement it with `std::thread` when we're using `std` and otherwise add an impl with FreeRTOS?

Reply:
As in implement it with `std::thread` when we are using `std` and *don't provide an implementation at all* for `no_std`. With `no_std`, user would be on their own to provide an implementation in whatever way they feel like.

It boils down to the following: *we have to keep the `esp-idf-svc` surface small, clean and easy to explain by exposing Rust-safe, self-contained pieces of functionality, - ideally - behind generic traits*. The emphasis being on safe and self-contained. We cannot just go there and expose random bits and pieces of all the raw level unsafe APIs that are out there in the ESP-IDF. For example, in your "basic freertos task api" what you have exposed from FreeRTOS is just a way to start a task (and that is always pinned to a core, if I'm not mistaken). Is that API generic enough so that folks can use it without calling other, unsafe Freertos APIs? Well, no.  For one, the queueing and synchronization primitives of freertos are not exposed as safe APIs, because you did not need them for your particular case. But very likely people *will* need these to do something meaningful with freertos. So, how are we evolving your Freertos code? Is it going to - at some point in time - reach a feature parity with the unsafe freertos APIs? 
* If yes, then we are basically creating a whole Freertos wrapper, and shouldn't we then instead work towards making [some of the Freertos Rust wrappers interop-ing with `esp-idf-svc`](https://github.com/lobaro/FreeRTOS-rust) interop-ing with `esp-idf-sys`?
 * If no, aren't we in a better position by saying to folks that there are NO freertos safe APIs, and they should just use the unsafe bindings which are out there in `esp-idf-sys`?